runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -0,0 +1,229 @@
+        AbstractTransactionalContext context = TransactionalContext.getCurrentContext();
+        long txBeginGlobalAddress = context.getSnapshotTimestamp();
+
+        this.mdKV.put(CheckpointEntry.CheckpointDictKey.START_TIME, startTime.toString());
+        this.mdKV.put(CheckpointEntry.CheckpointDictKey.START_LOG_ADDRESS, Long.toString(txBeginGlobalAddress));
+
+        ImmutableMap<CheckpointEntry.CheckpointDictKey,String> mdKV = ImmutableMap.copyOf(this.mdKV);
+        CheckpointEntry cp = new CheckpointEntry(CheckpointEntry.CheckpointEntryType.START,
+                author, checkpointID, mdKV, null);
+        startAddress = sv.append(Collections.singleton(streamID), cp, null);
+        postAppendFunc.accept(cp, startAddress);
+        return startAddress;
+    }
+
+    /** Append zero or more CONTINUATION records to this
+     *  object's stream.  Each will contain a fraction of
+     *  the state of the object that we're checkpointing
+     *  (up to batchSize items at a time).
+     *
+     *  Corfu client transaction management, if desired, is the
+     *  caller's responsibility.
+     *
+     *  The Iterators class appears to preserve the laziness
+     *  of Stream processing; we don't wish to use more
+     *  memory than strictly necessary to generate the
+     *  checkpoint.  NOTE: It would be even more useful if
+     *  the map had a lazy iterator: the eagerness of
+     *  map.keySet().stream() is not ideal, but at least
+     *  it should be much smaller than the entire map.
+     *
+     *  NOTE: The postAppendFunc lambda is executed in the
+     *  current thread context, i.e., inside of a Corfu
+     *  transaction, and that transaction will be *aborted*
+     *  at the end of this function.  Any Corfu data
+     *  modifying ops will be undone by the TXAbort().
+     *
+     * @return Stream of global log addresses of the
+     * CONTINUATION records written.
+     */
+    public List<Long> appendObjectState() {
+        ImmutableMap<CheckpointEntry.CheckpointDictKey,String> mdKV = ImmutableMap.copyOf(this.mdKV);
+        List<Long> continuationAddresses = new ArrayList<>();
+
+        Iterators.partition(map.keySet().stream()
+                .map(k -> {
+                    return new SMREntry("put",
+                            new Object[]{keyMutator.apply(k), valueMutator.apply(map.get(k))},
+                            Serializers.JSON);
+                }).iterator(), batchSize)
+                .forEachRemaining(entries -> {
+                    // Convert Object[] to SMREntry[], combined with byte count.
+                    // java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [Lorg.corfudb.protocols.logprotocol.SMREntry;
+                    int numEntries = ((List) entries).size();
+                    SMREntry e[] = new SMREntry[numEntries];
+                    for (int i = 0; i < numEntries; i++) {
+                        e[i] = (SMREntry) ((List) entries).get(i);
+                    }
```

Comment:
I feel like this should turn into putAll and also code could be java-fied down the line; but let's not wait for this, just remember to get back to it.